### PR TITLE
Add stack-trace docs

### PR DESCRIPTION
**Pull request by gerth2**
First pass at stack trace documentation.

Java-only for now, but should be easy to add C++. Soliciting feedback on tone and teaching technique first, then I'll go back to add the C++ side.

Should start to address https://github.com/wpilibsuite/frc-docs/issues/325

### CODE REVIEW

**Comment by jasondaming** [comment]
I like the content of this article.  I think it should go under Basic Programming not a new section.

Lots of linting errors I assume you just have your editor setup with spaces and CR which linting doesn't like.

File: source/docs/software/troubleshooting/robots-shouldnt-quit.rst [hunk 1]
```diff
@@ -0,0 +1,415 @@
+Robots Should Not Quit, But Yours Did!
+======================================
+
+``Robots should not quit, but yours did!``
+
+When your robot code hits an unexpected error, you will see this message show up in some console output (Driver Station or rioLog). You'll probably also notice your robot abruptly stop, or possibly never move. These unexpected errors are called *unhandled exceptions*.
```
Comment (CodingKoopa) [inline source/docs/software/troubleshooting/robots-shouldnt-quit.rst:6]: I don't think this capitalization of RioLog is correct.
image
```suggestion
When your robot code hits an unexpected error, you will see this message show up in some console output (Driver Station or RioLog). You'll probably also notice your robot abruptly stop, or possibly never move. These unexpected errors are called *unhandled exceptions*.

```

File: source/docs/software/troubleshooting/robots-shouldnt-quit.rst [hunk 2]
```diff
@@ -0,0 +1,415 @@
...
+This article will explore some of the tools and techniques involved in finding and fixing those bugs.
+
+What's a Stack Trace?
+---------------------
+
+The ``Robots should not quit`` message is a signal that a *stack trace* has been printed out. 
+
+In C++ and Java, a `stack <https://en.wikipedia.org/wiki/Call_stack>`_ data structure is used to store information about which function or method is currently being executed.
+
+A *stack trace* prints information about what was on this stack when the unhandled exception occurred.This points you to the lines of code which were running just before the problem happened. While it doesn't always point you to the *exact root cause* of your issue, it's usually the best place to start looking.
```
Comment (CodingKoopa) [inline source/docs/software/troubleshooting/robots-shouldnt-quit.rst:19]:
```suggestion
A *stack trace* prints information about what was on this stack when the unhandled exception occurred. This points you to the lines of code which were running just before the problem happened. While it doesn't always point you to the *exact root cause* of your issue, it's usually the best place to start looking.

```

Comment (CodingKoopa) [inline source/docs/software/troubleshooting/robots-shouldnt-quit.rst:19]: Is the emphasis on `exact root cause` necessary/consistent with the rest of the doc writing style?

File: source/docs/software/troubleshooting/robots-shouldnt-quit.rst [hunk 3]
```diff
@@ -0,0 +1,415 @@
...
+
+What's an "Unhandled Exception"?
```
Comment (CodingKoopa) [inline source/docs/software/troubleshooting/robots-shouldnt-quit.rst:21]: Quotes here aren't consistent with the other header.

File: source/docs/software/troubleshooting/robots-shouldnt-quit.rst [hunk 4]
```diff
@@ -0,0 +1,415 @@
...
+
+ * Access to your source code, (EX: :ref:`on github.com <docs/software/basic-programming/git-getting-started:Git Version Control Introduction>`)
+ * The **full text** of the error, including the full stack trace.
+
+Common Examples & Patterns
+--------------------------
+
+There are a number of common issues which result in runtime exceptions. 
+
+``Null``/``NULL``
```
Comment (CodingKoopa) [inline source/docs/software/troubleshooting/robots-shouldnt-quit.rst:130]: If possible, this might be better as:
```suggestion
``null``/``NULL``
```

File: source/docs/software/troubleshooting/robots-shouldnt-quit.rst [hunk 5]
```diff
@@ -0,0 +1,415 @@
...
+^^^^^^^^^^^^^^^^^
+
+Both C++ and Java have the concept of "null" - a reference which has not yet been initialized, and does not refer to anything meaningful.
```
Comment (CodingKoopa) [inline source/docs/software/troubleshooting/robots-shouldnt-quit.rst:133]: I'm not too keen on this explanation, since it might be the programmer's first introduction to `null`, which this doesn't give a broad enough description of IMO. Perhaps:
```suggestion
Both C++ and Java have the concept of "null", a value that does not refer to anything meaningful. In Java, `null` is assigned when there is no value given to a variable declared in a method, or a non-primitive (e.g. not an instance of a class) member of a class. In C++, TODO
```

File: source/docs/software/troubleshooting/robots-shouldnt-quit.rst [hunk 6]
```diff
@@ -0,0 +1,415 @@
...
+
+      .. code-block:: C++
+
+         //TODO
+
+
+Divide by Zero
+^^^^^^^^^^^^^^
+
+It is not generally possible to divide an integer by zero, and expect reasonable results. Most processors (including the RoboRIO) will cause an Unhandled Exception.
```
Comment (CodingKoopa) [inline source/docs/software/troubleshooting/robots-shouldnt-quit.rst:226]:
```suggestion
It is not generally possible to divide an integer by zero, and expect reasonable results. Most processors (including the roboRIO) will cause an Unhandled Exception.
```

Comment (CodingKoopa) [inline source/docs/software/troubleshooting/robots-shouldnt-quit.rst:226]: As far as wording goes, I say the processor itself isn't causing the exception.
```suggestion
It is not generally possible to divide an integer by zero, and expect reasonable results. Most processors (including the RoboRIO) will raise an Unhandled Exception.
```

File: source/docs/software/troubleshooting/robots-shouldnt-quit.rst [hunk 7]
```diff
@@ -0,0 +1,415 @@
...
+
+      .. code-block:: text
+
+          TODO
+
+Looking at the stack trace, we can see a ``java.lang.ArithmeticException: / by zero`` exception has occurred on line 25. If you look at the two variables which are used on the right-hand side of the ``=`` operator, you might notice one of them has not been initialized. This means its value is, by default, zero. And, the zero-value variable is used in the denominator of a division operation. Hence, the divide by zero error happens.
+
+Alternatively, by running the single-step debugger and stopping on line 25, you could inspect the value of all variables to discover ``shoulderToElbow_in`` has a value of ``0``.
+
+Fixing Div/0 Issues
```
Comment (CodingKoopa) [inline source/docs/software/troubleshooting/robots-shouldnt-quit.rst:283]: Unnecessary abbreviation?

File: source/docs/software/troubleshooting/robots-shouldnt-quit.rst [hunk 8]
```diff
@@ -0,0 +1,415 @@
...
+
+      .. code-block:: text
+
+          TODO
+
+This stack trace shows that a``edu.wpi.first.hal.util.UncleanStatusException`` has occurred. It also gives the helpful message: ``HAL: Resource already allocated``.
+
+Looking at our stack trace, we see that the error *actually* happened deep within some WPILib content. However, we should start by looking in our own code. Halfway through the stack trace, you can find a reference to the last line of the team's robot code that called into WPILib: ``Robot.java:26``.
+
+Taking a peek at the code, we see line 26 is where the second motor controller is declared. We can also note that *both* motor controllers are assigned to PWM output ``0``. This doesn't make logical sense, and isn't physically possible. Therefor, the WPILib libraries purposefully generate a custom error message and exception to alert the software developers of a non-achievable hardware configuration.
```
Comment (CodingKoopa) [inline source/docs/software/troubleshooting/robots-shouldnt-quit.rst:386]:
```suggestion
Taking a peek at the code, we see line 26 is where the second motor controller is declared. We can also note that *both* motor controllers are assigned to PWM output ``0``. This doesn't make logical sense, and isn't physically possible. Therefore, the WPILib libraries purposefully generate a custom error message and exception to alert the software developers of a non-achievable hardware configuration.
```

Comment (CodingKoopa) [inline source/docs/software/troubleshooting/robots-shouldnt-quit.rst:386]:
```suggestion
Taking a peek at the code, we see line 26 is where the second motor controller is declared. We can also note that *both* motor controllers are assigned to PWM output ``0``. This doesn't make logical sense, and isn't physically possible. Therefor, WPILib purposefully generates a custom error message and exception to alert the software developers of a non-achievable hardware configuration.
```

File: source/docs/software/troubleshooting/robots-shouldnt-quit.rst [hunk 9]
```diff
@@ -0,0 +1,415 @@
...
+
+   .. group-tab:: C++
+
+      .. code-block:: text
+
+          TODO
+
+This stack trace shows that a``edu.wpi.first.hal.util.UncleanStatusException`` has occurred. It also gives the helpful message: ``HAL: Resource already allocated``.
+
+Looking at our stack trace, we see that the error *actually* happened deep within some WPILib content. However, we should start by looking in our own code. Halfway through the stack trace, you can find a reference to the last line of the team's robot code that called into WPILib: ``Robot.java:26``.
```
Comment (CodingKoopa) [inline source/docs/software/troubleshooting/robots-shouldnt-quit.rst:384]: For brevity.
```suggestion
Looking at our stack trace, we see that the error *actually* happened deep within WPILib. However, we should start by looking in our own code. Halfway through the stack trace, you can find a reference to the last line of the team's robot code that called into WPILib: ``Robot.java:26``.
```

**Review by CodingKoopa — CHANGES_REQUESTED** [review]
This looks like an amazing resource for teams!

File: source/docs/software/troubleshooting/robots-shouldnt-quit.rst (hunk 4 above)
Comment (gerth2) [reply source/docs/software/troubleshooting/robots-shouldnt-quit.rst:132]: Lemme go look, this might get resolved as part of the C++ expansion. You got it though, the intent was to cover Java's `null` and C++'s `NULL`. I'm thinking an `inline code block containing NULL` will indicate the programming-language specific identifier, while plain old "null" will indicate the general concept.

File: source/docs/software/troubleshooting/robots-shouldnt-quit.rst (hunk 6 above)
Comment (gerth2) [reply source/docs/software/troubleshooting/robots-shouldnt-quit.rst:228]: Agreed, "Raise" is better here. 

I'll do some thinking as well as we start to reword for C++ too, as the core concept is straighforward, but nuanced:

The _problem_ is some logical inconsistency in the user's code, which puts the processor (or VM) in a state where it doesn't make sense to continue executing.
The exception users observe is a _result_ of the processor, the VM, or some code going into "we have a problem" mode, stopping the normal flow of execution, and raising the exception up.

File: source/docs/software/troubleshooting/robots-shouldnt-quit.rst (hunk 7 above)
Comment (gerth2) [reply source/docs/software/troubleshooting/robots-shouldnt-quit.rst:285]: Possibly. Lemme go check the exact text of what gets printed out in C++/Java statements. The only goal of using an alternate way to "spell" the same concept is to enhance searchability.

File: source/docs/software/troubleshooting/robots-shouldnt-quit.rst (hunk 2 above)
Comment (gerth2) [reply source/docs/software/troubleshooting/robots-shouldnt-quit.rst]: I'll got take a peek at the style guide, I'm not entirely sure. I'm definitely not tied to having it there, it's more a reflection of where I'd place my verbal emphasis if I were saying it out loud. 

My end goal would really just be to drive home the main point of the whole article: a stack trace is a _tool to solve your problem_, not a solution. I've found new folks often get confused (and discouraged) when they see an experienced person look at a stack trace (or similar), come up with an answer, and assume that the answer must have come from within said stack trace.

TL;DR probably doesn't have to be there.

**Comment by gerth2** [comment]
Ok. Getting close. C++ is the next big thing. Soliciting some help/ideas for the HAL resource already allocated.... In C++, I'm seeing just an "error", not a program crash or stack trace:

```
********** Robot program starting **********
Not loading CameraServerShared
Error at frc::PWM::PWM [PWM.cpp:32]: HAL: Resource already allocated, Minimum Value: 0, MaximumValue: 20, Requested Value: 0
```

I assume this is expected.... should the C++ section just skip over this?

File: source/docs/software/troubleshooting/robots-shouldnt-quit.rst (hunk 2 above)
Comment (CodingKoopa) [reply source/docs/software/troubleshooting/robots-shouldnt-quit.rst]: Okay, fair enough :)

File: source/docs/software/troubleshooting/robots-shouldnt-quit.rst (hunk 4 above)
Comment (CodingKoopa) [reply source/docs/software/troubleshooting/robots-shouldnt-quit.rst]: Ah, okay. In any case, if you want to talk about the concept of null values, that wouldn't be in a code tag, e.g.
> Null/`null`/`NULL`
although I personally would prefer just `null`/`NULL`.

File: source/docs/software/troubleshooting/robots-shouldnt-quit.rst (hunk 7 above)
Comment (CodingKoopa) [reply source/docs/software/troubleshooting/robots-shouldnt-quit.rst]: Ah okay. If it is using the same phrasing as the error message, then maybe consider putting the "Div/0" in a code block or italics.

**Comment by gerth2** [comment]
Putting back up for review. I think I hit all the comments and requests in the PR so far. I've taken a first-stab at C++, pushing users to go looking in VS Code for similar stack trace information in the Call Stack window. Let me know thoughts here or on discord!

File: source/docs/software/basic-programming/reading-stacktraces.rst [hunk 10]
```diff
@@ -0,0 +1,553 @@
...
+            public void robotInit() {
+
+               leftFrontMotor = new PWMSparkMax(0);
+               leftRearMotor = new PWMSparkMax(1);
+
+            }
+
+   .. group-tab:: C++
+
+      .. code-block:: C++
```
Comment (Daltz333) [inline source/docs/software/basic-programming/reading-stacktraces.rst:538]:
```suggestion
      .. code-block:: C++
```

**Comment by PeterJohnson** [comment]
The C++ stack traces thing just got changed in the 2022 branch.  See wpilibsuite/allwpilib#3306 and wpilibsuite/allwpilib#3305.

**Comment by gerth2** [comment]
> C++ stack traces just got changed/improved in the 2022 branch 

Sweet, just took a peek. The updates look pretty awesome.

Given the way the C++ content got written, it's definitely scoped to 2021 functionality. I don't currently have a full build environment for the 2022 branch, I can look into that.

The biggest thing that's going to be "wrong" going into 2022 is the note about needing the debugger to get the stack trace. Additionally, screenshots in the C++ tabs should probably get replaced with representative text from messages printed when the faulty examples get run.

**Comment by Daltz333** [comment]
In Thad's PR for duplicate entries, we could probably add his examples to the docs here.

**Comment by gerth2** [comment]
Kk. Still no progress on building latest wpilib to generate sample stacktraces in C++, mostly because I have not had time. But, will poke around for said thad pr and look about including it.

**Comment by sciencewhiz** [comment]
> But, will poke around for said thad pr and look about including it.

https://github.com/wpilibsuite/allwpilib/pull/3322

**Comment by Daltz333** [comment]
@gerth2 I'm quite satisfied with this PR as-is. Would you like to wait to implement the new clearier error messages, or should this be done later?

**Comment by gerth2** [comment]
Yea, that's good by me - From me looking into the other PR's mentioned, none of the information in it is wrong, just "less than complete" for 2022.

I'll enter in an issue referencing the changes so someone remembers to go update it. Once some beta release of wpilib 2022 is out it would be pretty straightforward for someone to copy in the bad-code to a new robot project, and update the screenshots & C++ descriptions to move away from recommending the debugger.

File: source/docs/software/basic-programming/reading-stacktraces.rst [hunk 11]
```diff
@@ -0,0 +1,553 @@
...
+
+Common Examples & Patterns
+--------------------------
+
+There are a number of common issues which result in runtime exceptions.
+
+Null Pointers and References
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+Both C++ and Java have the concept of "null" - a reference which has not yet been initialized, and does not refer to anything meaningful.
```
Comment (prateekma) [inline source/docs/software/basic-programming/reading-stacktraces.rst:163]: In C++, references cannot be null, only pointers can.

File: source/docs/software/basic-programming/reading-stacktraces.rst [hunk 12]
```diff
@@ -0,0 +1,553 @@
...
+
+      .. code-block:: Java
+          :lineno-start: 19
+
+            PWMSparkMax armMotorCtrl;
+
+            @Override
+            public void robotInit() {
+
+                armMotorCtrl.setInverted(true);
```
Comment (prateekma) [inline source/docs/software/basic-programming/reading-stacktraces.rst:181]: Remove extra spaces around this line.

File: source/docs/software/basic-programming/reading-stacktraces.rst [hunk 13]
```diff
@@ -0,0 +1,553 @@
...
+
+         class Robot : public frc::TimedRobot {
+            public:
+               void RobotInit() override {
+                  motorRef->SetInverted(false);
+               }
+
+            private:
+               frc::PWMVictorSPX m_armMotor{0};
+               frc::PWMVictorSPX * motorRef;
```
Comment (prateekma) [inline source/docs/software/basic-programming/reading-stacktraces.rst:198]: Remove space between `PWMVictorSPX` and `*`.

File: source/docs/software/basic-programming/reading-stacktraces.rst [hunk 14]
```diff
@@ -0,0 +1,553 @@
...
+
+      By going to line 24, you can see there is only one thing which could be null - ``armMotorCtrl``. Looking further up, you can see that the ``armMotorCtrl`` object is declared, but never instantiated.
+
+      Alternatively, you can step through lines of code with the single step debugger, and stop when you hit line 24. Inspecting the ``armMotorCtrl`` object at that point would show that it is null.
+
+   .. group-tab:: C++
+
+      .. code-block:: text
+
+         Exception has occurred: W32/0xc0000005
```
Comment (prateekma) [inline source/docs/software/basic-programming/reading-stacktraces.rst:231]: Might be worth mentioning that this error will look different on the RIO / other non-Windows platforms.

File: source/docs/software/basic-programming/reading-stacktraces.rst [hunk 15]
```diff
@@ -0,0 +1,553 @@
...
+         this->motorRef was nullptr.
+
+      In Simulation, this will show up in a debugger window that points to line 20 in the above buggy code.
+
+      You can view the full stack trace by clicking the debugger tab in VS Code:
+
+      .. image:: images/reading-stacktraces/cpp_null_stacktrace.png
+         :alt: Stack Trace associated with a null-related error
+
+      The error is specific - our member variable ``motorRef`` was declared, but never assigned a value. Therefor, when we attempt to use it to call a method using the ``->`` operator, the exception occurs.
```
Comment (prateekma) [inline source/docs/software/basic-programming/reading-stacktraces.rst:242]: Check spelling of "Therefore".

File: source/docs/software/basic-programming/reading-stacktraces.rst [hunk 16]
```diff
@@ -0,0 +1,553 @@
...
+      .. code-block:: Java
+          :lineno-start: 19
+
+            PWMSparkMax armMotorCtrl;
+
+            @Override
+            public void robotInit() {
+
+                armMotorCtrl = new PWMSparkMax(0);
+                armMotorCtrl.setInverted(true);
```
Comment (prateekma) [inline source/docs/software/basic-programming/reading-stacktraces.rst:266]: Remove spaces around these lines.

File: source/docs/software/basic-programming/reading-stacktraces.rst [hunk 17]
```diff
@@ -0,0 +1,553 @@
...
+         class Robot : public frc::TimedRobot {
+            public:
+               void RobotInit() override {
+                  motorRef = &m_armMotor;
+                  motorRef->SetInverted(false);
+               }
+
+            private:
+               frc::PWMVictorSPX m_armMotor{0};
+               frc::PWMVictorSPX * motorRef;
```
Comment (prateekma) [inline source/docs/software/basic-programming/reading-stacktraces.rst:284]: See comment above about space between `*` and type.

File: source/docs/software/basic-programming/reading-stacktraces.rst [hunk 18]
```diff
@@ -0,0 +1,553 @@
...
+         };
+
+
+
+Divide by Zero
+^^^^^^^^^^^^^^
+
+It is not generally possible to divide an integer by zero, and expect reasonable results. Most processors (including the roboRIO) will raise an Unhandled Exception.
```
Comment (prateekma) [inline source/docs/software/basic-programming/reading-stacktraces.rst:292]: "generally not possible" sounds better to me. What do you think?

File: source/docs/software/basic-programming/reading-stacktraces.rst [hunk 19]
```diff
@@ -0,0 +1,553 @@
...
+          :lineno-start: 18
+
+            int armLengthRatio;
+            int elbowToWrist_in = 39;
+            int shoulderToElbow_in;
+
+            @Override
+            public void robotInit() {
+
+               armLengthRatio = elbowToWrist_in / shoulderToElbow_in;
```
Comment (prateekma) [inline source/docs/software/basic-programming/reading-stacktraces.rst:310]: Remove blank lines around this line.

File: source/docs/software/basic-programming/reading-stacktraces.rst [hunk 20]
```diff
@@ -0,0 +1,553 @@
...
+For example, consider the following code:
+
+.. tabs::
+
+   .. group-tab:: Java
+
+      .. code-block:: Java
+          :lineno-start: 18
+
+            int armLengthRatio;
```
Comment (prateekma) [inline source/docs/software/basic-programming/reading-stacktraces.rst:303]: This behavior of ints being initialized to zero is not exactly clear so I think we should explicitly initialize it.

File: source/docs/software/basic-programming/reading-stacktraces.rst [hunk 21]
```diff
@@ -0,0 +1,553 @@
...
+          :lineno-start: 17
+
+            class Robot : public frc::TimedRobot {
+               public:
+               void RobotInit() override {
+                  armLengthRatio = elbowToWrist_in / shoulderToElbow_in;
+               }
+
+               private:
+                  int armLengthRatio;
```
Comment (prateekma) [inline source/docs/software/basic-programming/reading-stacktraces.rst:326]: In C++, this needs to be explicitly set to 0; otherwise it is UB. (Windows will sometimes not zero-initialize whereas Linux and macOS will).

File: source/docs/software/basic-programming/reading-stacktraces.rst [hunk 22]
```diff
@@ -0,0 +1,553 @@
...
+
+      Alternatively, by running the single-step debugger and stopping on line 25, you could inspect the value of all variables to discover ``shoulderToElbow_in`` has a value of ``0``.
+
+   .. group-tab:: C++
+
+
+      .. code-block:: text
+
+         Exception has occurred: W32/0xc0000094
+         Unhandled exception at 0x00007FF71B223CD6 in frcUserProgram.exe: 0xC0000094: Integer division by zero.
```
Comment (prateekma) [inline source/docs/software/basic-programming/reading-stacktraces.rst:361]: See comment about C++ exception message earlier.

File: source/docs/software/basic-programming/reading-stacktraces.rst [hunk 23]
```diff
@@ -0,0 +1,553 @@
...
+               at frc.robot.Robot.robotInit(Robot.java:25)
+               at edu.wpi.first.wpilibj.TimedRobot.startCompetition(TimedRobot.java:94)
+               at edu.wpi.first.wpilibj.RobotBase.runRobot(RobotBase.java:335)
+               at edu.wpi.first.wpilibj.RobotBase.lambda$startRobot$0(RobotBase.java:387)
+               at java.base/java.lang.Thread.run(Thread.java:834)
+
+         Warning at edu.wpi.first.wpilibj.RobotBase.runRobot(RobotBase.java:350): unexpected error has occurred, but yours did!
+         Error at edu.wpi.first.wpilibj.RobotBase.runRobot(RobotBase.java:352): The startCompetition() method (or methods called by it) should have handled the exception above.
+
+      Looking at the stack trace, we can see a ``java.lang.ArithmeticException: / by zero`` exception has occurred on line 25. If you look at the two variables which are used on the right-hand side of the ``=`` operator, you might notice one of them has not been initialized. This means its value is, by default, zero. And, the zero-value variable is used in the denominator of a division operation. Hence, the divide by zero error happens.
```
Comment (prateekma) [inline source/docs/software/basic-programming/reading-stacktraces.rst:351]: I see here that you mention uninitialized ints will default to zero, so maybe you can disregard the comment above in Java. C++ comment still applies however.

File: source/docs/software/basic-programming/reading-stacktraces.rst [hunk 24]
```diff
@@ -0,0 +1,553 @@
...
+          :lineno-start: 18
+
+            int armLengthRatio;
+            int elbowToWrist_in = 39;
+            int shoulderToElbow_in = 3;
+
+            @Override
+            public void robotInit() {
+
+               armLengthRatio = elbowToWrist_in / shoulderToElbow_in;
```
Comment (prateekma) [inline source/docs/software/basic-programming/reading-stacktraces.rst:397]: Remove blank lines around this.

File: source/docs/software/basic-programming/reading-stacktraces.rst [hunk 25]
```diff
@@ -0,0 +1,553 @@
...
+            class Robot : public frc::TimedRobot {
+               public:
+               void RobotInit() override {
+                  armLengthRatio = elbowToWrist_in / shoulderToElbow_in;
+               }
+
+               private:
+                  int armLengthRatio;
+                  int elbowToWrist_in = 39;
+                  int shoulderToElbow_in; = 3
```
Comment (prateekma) [inline source/docs/software/basic-programming/reading-stacktraces.rst:416]: Check semicolon placement.

**Review by prateekma — CHANGES_REQUESTED** [review]

File: source/docs/software/basic-programming/reading-stacktraces.rst (hunk 23 above)
Comment (gerth2) [reply source/docs/software/basic-programming/reading-stacktraces.rst:351]: I went ahead and changed the error in coding to be something like:

```
variable = 0; //TODO

```

to imply the situation where someone put it there as placeholder but forgot to change it. That should hopefully sidestep the init question (as I believe the answer changes for java/c++, as well as windows/linux and stack vs. heap?)

File: source/docs/software/basic-programming/reading-stacktraces.rst (hunk 11 above)
Comment (gerth2) [reply source/docs/software/basic-programming/reading-stacktraces.rst:163]: cranked the hand-waving up to 11 to sidestep this.

**Review by prateekma — APPROVED** [review]

**Review by Daltz333 — APPROVED** [review]
